File: khet-game-server/src/main/scala/com/github/mich8bsp/Move.scala
package com.github.mich8bsp

trait Move {
  def reverse(): Move
  def moveType: EMoveType
}

case class RotationMove( pos: BoardPos,  direction: ERotationDirection) extends Move {
  override def reverse(): Move = RotationMove(pos.reverse(), direction)
  override def moveType: EMoveType = EMoveType.ROTATE
}
case class PositionMove( from: BoardPos,  to: BoardPos) extends Move {
  override def reverse(): Move = PositionMove(from.reverse(), to.reverse())
  override def moveType: EMoveType = EMoveType.POSITION
}
case class SwitchMove( pos1: BoardPos,  pos2: BoardPos) extends Move {
  override def reverse(): Move = SwitchMove(pos1.reverse(), pos2.reverse())
  override def moveType: EMoveType = EMoveType.SWITCH
}

case class BoardPos(i: Int, j: Int){
  def reverse(): BoardPos = BoardPos(GameConfig.ROWS-i-1, GameConfig.COLS-j-1)
}

File: khet/core/src/com/github/mich8bsp/logic/Move.kt
package com.github.mich8bsp.logic

interface Move{
    fun getMoveType(): EMoveType
}

data class RotationMove(val pos: BoardPos, val direction: ERotationDirection) : Move {
    override fun getMoveType(): EMoveType = EMoveType.ROTATE
}

data class PositionMove(val from: BoardPos, val to: BoardPos): Move {
    override fun getMoveType(): EMoveType = EMoveType.POSITION
}
data class SwitchMove(val pos1: BoardPos, val pos2: BoardPos): Move {
    override fun getMoveType(): EMoveType = EMoveType.SWITCH
}

data class MoveRecord<T : Move>(val move: T, val playerColor: EPlayerColor, val recordId: Long)